src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -392,15 +392,21 @@ public static AbstractMemorySegmentImpl ofBuffer(ByteBuffer bb) {
 
         int pos = bb.position();
         int limit = bb.limit();
-
-        MemoryScope bufferScope = new MemoryScope(bb, null);
         int size = limit - pos;
 
-        MemorySegment bufferSegment = (MemorySegment)nioAccess.bufferSegment(bb);
-        int modes = bufferSegment == null ?
-                defaultAccessModes(size) : bufferSegment.accessModes();
-        Thread owner = bufferSegment == null ?
-                Thread.currentThread() : bufferSegment.ownerThread();
+        AbstractMemorySegmentImpl bufferSegment = (AbstractMemorySegmentImpl)nioAccess.bufferSegment(bb);
```

Comment:
Wondering if we can't just return `bufferSegment` here? It should have the write type right? I think the only thing we'd need to do is drop the WRITE access mode if `bb` is readOnly.

Reply:
uhmmm - not quite - the buffer could have been sliced?

> Wrt your concerns. I think we need to preserve the scope to prevent dead accesses. Consider this example:
> 
> ```
>         MemorySegment s1 = MemorySegment.allocateNative(intArrayLayout);
>         MemorySegment s2 = MemorySegment.ofByteBuffer(s1.asByteBuffer());
> 
>         s1.close(); // memory freed
> 
>         intElemHandle.set(s2.baseAddress(), (long) 0, 10); // Dead access!
> ```
> 
> It fails with this patch, but wrongly succeeds without it.

Very good point. Aliasing is an issue w.r.t. closing.